### PR TITLE
feat(puffin): add file writer

### PR DESCRIPTION
**Pull request by zhongzc**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

Add writer for Puffin in sync and async manner.

Mainly introduce `PuffinSyncWriter` and `PuffinAsyncWriter` traits and implementations.

Please note that `src/puffin/src/tests.rs` is not a bin file, check it by clicking View File.

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)
#2705

### CODE REVIEW

File: src/puffin/src/tests.rs
Comment (killme2008) [inline src/puffin/src/tests.rs:1]: Is it a binary file?

File: src/puffin/src/file_format/writer/file.rs [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+        let size = io::copy(&mut blob.data, &mut self.writer).context(WriteSnafu)?;
+
+        let blob_metadata = self.create_blob_metadata(blob.blob_type, blob.properties, size);
+        self.blob_metadata.push(blob_metadata);
+
+        self.next_blob_offset += size;
+        Ok(())
+    }
+
+    fn finish(&mut self) -> Result<()> {
```
Comment (killme2008) [inline src/puffin/src/file_format/writer/file.rs:87]: Non-blocking:  is it necessary to call the `finish` when `drop` the struct automatically?

File: src/puffin/src/tests.rs
Comment (zhongzc) [reply src/puffin/src/tests.rs:1]: No, it seems to be a GitHub bug. You can use View File to see the content.

File: src/puffin/src/file_format/writer/footer.rs [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    pub fn new(blob_metadata: Vec<BlobMetadata>, file_properties: HashMap<String, String>) -> Self {
+        Self {
+            blob_metadata,
+            file_properties,
+        }
+    }
+
+    /// Serializes the footer to bytes
+    pub fn into_footer_bytes(mut self) -> Result<Vec<u8>> {
+        let mut buf = Vec::new();
```
[the remark this answers is not in the document]
Comment (zhongzc) [reply src/puffin/src/file_format/writer/footer.rs]: updated

File: src/puffin/src/file_format/writer/file.rs (hunk 1 above)
Comment (zhongzc) [reply src/puffin/src/file_format/writer/file.rs:86]: it's really hard to do async drop in Rust

**Review by killme2008 — APPROVED** [review]
LGTM

File: src/puffin/src/file_format/writer/file.rs [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+        let blob_metadata = self.create_blob_metadata(blob.blob_type, blob.properties, size);
+        self.blob_metadata.push(blob_metadata);
+
+        self.next_blob_offset += size;
+        Ok(())
+    }
+
+    async fn finish(&mut self) -> Result<()> {
+        self.write_header_if_needed_async().await?;
+        self.write_footer_async().await
```
Comment (evenyag) [inline src/puffin/src/file_format/writer/file.rs:114]: We might need to call `flush()` explicitly to ensure data is flushed, especially for async writers as they can't flush on drop.

Comment (zhongzc) [reply src/puffin/src/file_format/writer/file.rs]: updated

File: src/puffin/src/file_format.rs [hunk 4]
```diff
@@ -57,6 +57,8 @@ pub const MIN_FOOTER_SIZE: u64 = MAGIC_SIZE * 2 + FLAGS_SIZE + PAYLOAD_SIZE_SIZE
 bitflags! {
     #[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
     pub struct Flags: u32 {
+        const DEFAULT = 0b00000000;
```
Comment (killme2008) [inline src/puffin/src/file_format.rs:60]: 👍🏻.  This is my point.

File: src/puffin/src/file_format/writer/file.rs [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+        self.blob_metadata.push(blob_metadata);
+
+        self.next_blob_offset += size;
+        Ok(())
+    }
+
+    async fn finish(&mut self) -> Result<()> {
+        self.write_header_if_needed_async().await?;
+        self.write_footer_async().await?;
+        self.writer.flush().await.context(FlushSnafu)
```
Comment (evenyag) [inline src/puffin/src/file_format/writer/file.rs:116]: I realize that some `flush()` implementations may only write data to the underlying `writer`. We also need to call `close()` to really flush and close the `writer`.

https://docs.rs/futures/latest/futures/io/trait.AsyncWriteExt.html#method.close

**Review by evenyag — APPROVED** [review]
Rest LGTM. You can merge the latest develop branch to pass the codecov check.

**Pull request merged**